File: 1013/Main.java
import java.util.Scanner;

public class Main {
	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		int N = scanner.nextInt();
		int M = scanner.nextInt();
		int K = scanner.nextInt();
		int[][] graph = new int[N+1][N+1];
		for (int i = 0; i <= N; i++) {
			for (int j = 0; j <= N; j++) {
				graph[i][j] = 0;
			}
		}
		for (int i = 0; i < M; i++) {
			int s = scanner.nextInt();
			int t = scanner.nextInt();
			graph[s][t] = 1;
			graph[t][s] = 1;
		}
		int[] concern = new int[K];
		for (int i = 0; i < K; i++) {
			concern[i] = scanner.nextInt();
		}
		for (int c : concern) {
			int[][] g = new int[N+1][N+1];
			for (int i = 0; i <= N; i++) {
				for (int j = 0; j <= N; j++) {
					g[i][j] = graph[i][j];
				}
			}
			for (int i = 0; i <= N; i++) {
				g[c][i] = 0;
				g[i][c] = 0;
			}
			
			int[] visit = new int[N+1];
			for (int i = 0; i <= N; i++) {
				visit[i] = 0;
			}
			int component = 0;
			while (hasUnvisited(visit, N)) {
				for (int i = 1; i <= N; i++) {
					if (visit[i] == 0) {
						component++;
						dfs(i, g, visit, N);
						break;
					}
				}
			}
			System.out.println(component - 1 - 1);
		}
	}
	
	public static void dfs(int v, int[][] g, int[] visit, int N) {
		visit[v] = 1;
		for (int i = 1; i <= N; i++) {
			if (g[v][i] == 1 && visit[i] == 0) {
				dfs(i, g, visit, N);
			}
		}
	}
	
	public static boolean hasUnvisited(int[] visit, int N) {
		for (int i = 1; i <= N; i++) {
			if (visit[i] == 0)
				return true;
		}
		return false;
	}
}

File: 1013/main.cpp
#include <stdio.h>
#include <iostream>
#include <string.h>

using namespace std;

int graph[1002][1002];
int N, M, K;
int visit[1002];

void dfs(int v);

int main()
{
    memset(graph, 0, sizeof(graph));// memset 以字节为单位，对 int 只能批量初始化0、-1，要初始化为其他值，参考fill、fill_n
    // cin、cout 会超时，scanf、printf 不会超时
    // cin >> N >> M >> K;
    scanf("%d %d %d", &N, &M, &K);
    for (int i = 0; i < M; i++)
    {
        int j, k;
        // cin >> j >> k;
        scanf("%d %d", &j, &k);
        graph[j][k] = 1;
        graph[k][j] = 1;
    }
    int query[K];
    for (int i = 0; i < K; i++)
        // cin >> query[i];
        scanf("%d", &query[i]);

    for (int i = 0; i < K; i++)
    {
        int d = query[i];
        fill_n(visit, 1002, 0);// 批量置 0
        visit[d] = 1;
        int component = 0;
        for (int j = 1; j <= N; j++)
        {
            if (visit[j] == 0)
            {
                component++;
                dfs(j);
            }
        }
        // cout << component - 1 << endl;
        printf("%d\n", component - 1);
    }
    return 0;
}

// 注意数组传参
void dfs(int v)
{
    visit[v] = 1;
    for (int i = 1; i <= N; i++)
        if (graph[v][i] == 1 && visit[i] == 0)
            dfs(i);
}
